### DISCUSSION TITLE
Adonis drive getSignedUrl needs to be async?

### DISCUSSION THREAD

**Discussion by satheler** [adonisjs/core#3556] [General]
Here is a part of my code
```js
const path = seller.hasLogo ? seller.id : 'no-image'
seller.logo_url = await Drive.getSignedUrl(`${BASE_IMAGE_PATH}/${path}`, { contentType: 'image/jpeg' })
 ```

I use AWS S3 to store my files and before using Adonis Drive I had my own implementation. 
In my implementation I used this part inside a @computed property, in contrast, with Adonis Drive I can't because it's an asynchronous method.

I know that DriverContract needs to be abstract, but instead of return a `Promise<String>`, could to return `Promise<String> | String` and delegates to whoever is implementing it decides if their drive allows synchronous url generation ou not.

**Comment by thetutlage** [comment]
U mean you have a custom s3 driver that can compute the signedUrl synchronously?

**Reply by satheler** [reply]
Exactly. I used [aws-sdk](https://github.com/aws/aws-sdk-js) and in its type it's declared as a synchronous function.

<img width="907" alt="image" src="https://user-images.githubusercontent.com/6925493/153777576-4f5b3795-f64e-46d2-8100-616ebc76afad.png">

I was see how to do in AWS SDK v3 (modularized version) and is can't possible.

**Reply by satheler** [reply]
Maybe, using the custom signed url. Because, to create a signed url needs 5 informations. (1) bucket (2) key (3) region (4) access key id (5) secret access key.

With bucket, key and region we already have the S3 URL.
With this URL we can signed using SHA256 with access key id and secret access key.

**Reply by satheler** [reply]
In the new version, the AWS uses a asynchronous dependency to handle `sha256`.

**Reply by thetutlage** [reply]
Since you are using the aws-sdk directly. You can handle it outside of Drive as well.

**Reply by satheler** [reply]
Yep. That's was my solution 😄. Thank you.
